### PR TITLE
Suppress recurring errors when dropping unknown but active replication slots

### PR DESCRIPTION
**Pull request by mbanck**
When a replication slot is not registered with Patroni but is active, Patroni would log an error during each HA cycle in certain conditions (after a restart or role change). To avoid this, first check if the replication slot we are about to drop is still active and if so, only log a warning. Otherwise, log the slot we are dropping for informational purposes.

Close: #2499

The warning is only logged once according to my tests (unlike the error, which was carried over for each HA cycle).

### CODE REVIEW

File: patroni/postgresql/slots.py [hunk 1]
```diff
@@ -192,10 +192,19 @@ def ignore_replication_slot(self, cluster, name):
...
-                              'FROM pg_catalog.pg_replication_slots WHERE slot_name = %s AND NOT active)'), name, name)
-        # In normal situation rowcount should be 1, otherwise either slot doesn't exists or it is still active
-        return cursor.rowcount == 1
+        # First check if the replication slot is still active; not being able
+        # to drop it in this case is not considered an error
+        cursor_slot_active = self._query(('SELECT 1 FROM pg_catalog.pg_replication_slots WHERE ' +
+                                          'slot_name = %s AND active'), name)
+        if (cursor_slot_active.rowcount == 1):
+            logger.warning("Unable to drop unknown replication slot '%s', slot is still active", name)
+            return True
```
Comment (CyberDem0n) [inline patroni/postgresql/slots.py:201]: Since the method returns True from here there will be no further attempt to remove the slot.

Comment (mbanck) [reply patroni/postgresql/slots.py:201]: oh right I see - if the slots becomes inactive, we'd never drop it.

So let's get back to logging this at debug each cycle, will revise

Comment (mbanck) [reply patroni/postgresql/slots.py]: I rearranged this now, I hope this is better

File: patroni/postgresql/slots.py [hunk 2]
```diff
@@ -201,8 +201,17 @@ def _drop_incorrect_slots(self, cluster, slots, paused):
...
         for name in set(self._replication_slots) - set(slots):
             if not paused and not self.ignore_replication_slot(cluster, name) and not self.drop_replication_slot(name):
-                logger.error("Failed to drop replication slot '%s'", name)
                 self._schedule_load_slots = True
+                # Check if slot is still active, this is not considered an error
+                cursor = self._query(('SELECT 1 FROM pg_catalog.pg_replication_slots WHERE ' +
+                                      'slot_name = %s AND active'), name)
+                if (cursor.rowcount == 1):
+                    logger.debug("Unable to drop unknown replication slot '%s', slot is still active", name)
+                    return True
```
Comment (CyberDem0n) [inline patroni/postgresql/slots.py:210]: If we return from here it will prevent other slots from removal...

Comment (mbanck) [reply patroni/postgresql/slots.py]: Yeah, that was a copy-pasto I saw independently and have now removed

File: patroni/postgresql/slots.py [hunk 3]
```diff
@@ -201,8 +201,16 @@ def _drop_incorrect_slots(self, cluster, slots, paused):
...
                 self._schedule_load_slots = True
+                # Check if slot is still active, this is not considered an error
+                cursor = self._query(('SELECT 1 FROM pg_catalog.pg_replication_slots WHERE ' +
+                                      'slot_name = %s AND active'), name)
+                if (cursor.rowcount == 1):
+                    logger.debug("Unable to drop unknown replication slot '%s', slot is still active", name)
+                else:
+                    logger.error("Failed to drop replication slot '%s'", name)
+            else:
+                logger.info("Dropped unknown replication slot '%s'", name)
```
Comment (hughcapet) [inline patroni/postgresql/slots.py:213]: I'm afraid, this will be erroneously logged for 'paused' mode or an ignored slot as well

Comment (CyberDem0n) [reply patroni/postgresql/slots.py:213]: @mbanck please take a look

Comment (mbanck) [reply patroni/postgresql/slots.py]: You're right, that was a thinko; I've pushed a new version now

**Comment by hughcapet** [comment]
👍

**Comment by CyberDem0n** [comment]
:+1:

**Comment by CyberDem0n** [comment]
Merged, thank you @mbanck!

**Pull request merged**